### DISCUSSION TITLE
Dereference into an indexer

### DISCUSSION THREAD

**Discussion by Ai-N3rd** [dotnet/csharplang#8019] [Language Ideas]
### Problem:
Currently, when accessing an indexer on a pointer in C#, developers have to use explicit dereferencing syntax, which can be cumbersome and less intuitive. For instance:

```csharp
List<int> list = [ 1, 2, 3, 4 ];
List<int> *listPtr = &list;
int item = (*listPtr)[1];
```

### Proposal:
To improve readability and reduce cognitive load for developers working with pointers and indexers, I propose introducing a new syntactic sugar for pointer dereferencing with indexers.

New Syntax:

```csharp
Copy code
List<int> list = [ 1, 2, 3, 4 ];
List<int> *listPtr = &list;
int item = listPtr->[1];
```

**Comment by huoyaoyuan** [comment]
Pointer is much less used in C# comparing to C++. Pointer to `List<T>` or other reference types needs to be used with extra caution, or it would introduce GC hole.

And your syntax `&[ 1, 2, 3 ]` doesn't work - you can't get address of a literal expression. The compiler is free to store the instance at anywhere.
